README.md
# Timesheets

[![Build Status](https://travis-ci.org/demalexx/timesheets.svg?branch=master)](https://travis-ci.org/demalexx/timesheets)
[![codecov](https://codecov.io/gh/demalexx/timesheets/branch/master/graph/badge.svg)](https://codecov.io/gh/demalexx/timesheets)

Sublime Text 3 package to help working with IPONWEB timesheets (Jira).
Features:

* Syntax highlight;
* Open ticket under cursor in default browser;
* Help adding new timesheet lines using shortcut;
* Show how much time is worked today and this week in status bar.

![Screenshot](images/demo.png)

# Installation

Copy file [release/timesheets.sublime-package](https://github.com/demalexx/timesheets/raw/master/release/timesheets.sublime-package)
into Sublime package folder, depending on your OS.

*MacOS*

```
rm "$HOME/Library/Application Support/Sublime Text 3/Installed Packages/timesheets.sublime-package"
wget https://github.com/demalexx/timesheets/raw/master/release/timesheets.sublime-package -P "$HOME/Library/Application Support/Sublime Text 3/Installed Packages/"
```

*Ubuntu*

```
rm "$HOME/.config/sublime-text-3/Installed Packages/timesheets.sublime-package"
wget https://github.com/demalexx/timesheets/raw/master/release/timesheets.sublime-package -P "$HOME/.config/sublime-text-3/Installed Packages/"
```

*Windows*

```
Invoke-WebRequest -Uri https://github.com/demalexx/timesheets/raw/master/release/timesheets.sublime-package -OutFile "~/AppData/Roaming/Sublime Text 3/Installed Packages/timesheets.sublime-package"
```

# Usage

## Syntax highlight

To use syntax highlight select Timesheet from right bottom menu
of available syntaxes, or type "Set syntax: Timesheet"
in command palette (`Ctrl+Shift+P`, or `Cmd+Shift+P` on Mac).

If timesheet line has invalid format, it won't be highlighted
(e.g. missing comma, invalid date/time etc).
There is one exception - "time to" field could be filled with spaces
and line is highlighted as valid. Such line is considered as ticket 
you're working on right now, and finish time is not known yet.

## Open ticket in browser

If line contains valid ticket, you could open it in browser by:

* `Ctrl+Click` (`Alt+Click` on Mac) on any place on line;
* Select `Goto Ticket` in context menu.

## Shortcut to add timesheet lines

Shortcut `Alt+D` takes timesheet line under the cursor, 
clones it, modifies to reflect current date/time
and appends it to proper place. Also it fills empty time_to field.

It's easier to describe it through examples.

Suppose today is 2020-10-10 10:03 (`|` represents cursor position).

Before:
```
2020-10-01,10:00,12:00,PROJECT-123,working|
```

After:
```
2020-10-01,10:00,12:00,PROJECT-123,working

2020-10-10,10:00,     ,PROJECT-123,working|
```

- Blank line is inserted as separator between days;
- Original issue and comment is copied;
- Today is inserted;
- Current time is rounded and inserted as time_from;
- time_to field is left blank.

Suppose now is 11:08.

Before:
```
2020-10-01,10:00,12:00,PROJECT-123,"working"

2020-10-10,10:00,     ,PROJECT-123,"working|"
```

After:
```
2020-10-01,10:00,12:00,PROJECT-123,"working"

2020-10-10,10:00,11:00,PROJECT-123,"working"
2020-10-10,11:00,     ,PROJECT-123,"working|"
```

- Blank line is not inserted because it's same day;
- Original issue and comment is copied;
- Empty time_to is filled with rounded current time;
- time_to of original line is inserted as time_from;
- time_to field is left blank.

There is another shortcut `Alt+Shift+D` which does all the same
but doesn't copy issue and comment.

Before:
```
2018-01-01,10:00,12:00,|PROJECT-123,"working"
```

After:
```
2018-01-01,10:00,12:00,PROJECT-123,"working"

2018-01-10,10:00,     ,|,""
```

## Info about how much time is worked

Worked time today and this week is shown in status bar. 
It's updated automatically on tab activation or file save.

Examples: `Worked today 02:35, week 10:35` (hours:minutes).

If timesheet line has empty time_to field,
it's replaced with current time to calculate working time.

# Customization

Plugin has some settings accessible through
`Preferences → Package Settings → Timesheets → Settings`:

```javascript
{
    // Jira ticket URL, e.g.
    // "https://jira.example.com/browse/{}"
    "jira_ticket_url": "https://jira.iponweb.net/browse/{}"
}
```

Mouse bindings could be changed in
`Preferences → Package Settings → Timesheets → Mouse Bindings`:

Key bindings could be changed in
`Preferences → Package Settings → Timesheets → Key Bindings`:


CHANGELOG.md
# 1.0.0 (2020-10-26)

* Updated syntax highlight: 
  support comments without quotes, 
  support activity (DEV/MET/etc);
* Removed supporting RT;
* Removed build systems (checkout and commit, not applicable anymore).

# 0.5.0 (2018-07-14)

* Also show how much time is worked this week.

# 0.4.0 (2018-07-10)

* Added info about how much time is worked today into status bar;
* Fixed bug when SublimeLinter is installed
  (it crashed when timesheets plugin is installed);
* Added syntax and unit tests.

# 0.3.0 (2018-06-12)

* Added feature to insert timesheet lines
  based on already existing lines and fill time fields,
  using single shortcut (improved "duplicate line");
* Improved syntax highlight: now it doesn't highlight
  lines with invalid dates and times (e.g. "2018-13-35", "24:60").
  Still it's still just dumb regexp.

# 0.2.0 (2018-06-10)

* Added feature to checkout and commit timesheets
  directly from Sublime Text.
  Sublime's Build System is used for this.

# 0.1.0 (2018-06-02)

* Initial release.

tests/base.py
from unittest import TestCase

import sublime

from timesheets.helpers import SublimeHelper, TimesheetHelper


class BasePluginTestCase(TestCase):
    """
    Convenient class that creates view for tests and close it after test run.
    Also contains utility methods.
    """

    def setUp(self):
        # Create view that will be used by tests
        self.view = sublime.active_window().new_file()

        self.sublime_helper = SublimeHelper(self.view)
        self.timesheet_helper = TimesheetHelper(self.sublime_helper)

    def tearDown(self):
        if self.view:
            # Set scratch flag so view with content could be closed
            # without saving prompt
            self.view.set_scratch(True)
            self.view.window().run_command('close_file')

    def get_valid_line(self):
        valid_line = '2018-07-01,10:00,12:10,PROJECT-123,comment'
        self.assertTrue(
            self.timesheet_helper.extract_timesheet_info(valid_line)
        )
        return valid_line

    def get_invalid_line(self):
        invalid_line = '2018-07-01 10:00 12:10 PROJECT-123 comment'
        self.assertFalse(
            self.timesheet_helper.extract_timesheet_info(invalid_line)
        )
        return invalid_line

    def append_text(self, text: str):
        """Appends given text to view."""
        self.view.run_command('insert', {'characters': text})

    def get_text(self) -> str:
        """Return full content of current view."""
        return self.view.substr(sublime.Region(0, self.view.size()))

    def move_cursor(self, row: int, col: int):
        pos = self.view.text_point(row, col)
        self.view.sel().clear()
        self.view.sel().add(pos)


tests/__init__.py
import os
import sys

# In order to import modules from this package (like dateutil, freezegun),
# adjust path
sys.path.append(os.path.dirname(os.path.realpath(__file__)))


tests/test_helpers.py
from timesheets.tests.base import BasePluginTestCase
from timesheets.helpers import prettify_minutes


class TestTimesheetHelper(BasePluginTestCase):
    def test_prettify_minutes(self):
        self.assertEqual('01:05', prettify_minutes(65))


goto_ticket.py
import webbrowser

import sublime
import sublime_plugin

from timesheets.typing.typing import Optional
from timesheets.helpers import SublimeHelper, TimesheetHelper


class GotoTicketCommand(sublime_plugin.TextCommand):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.sublime_helper = SublimeHelper(self.view)
        self.timesheet_helper = TimesheetHelper(self.sublime_helper)

    def run(self, edit):
        """Entry-point, called when command is executed"""
        line_content = self.sublime_helper.get_current_line_content()
        timesheet_info = self.timesheet_helper.extract_timesheet_info(
            line_content
        )

        if timesheet_info:
            ticket_url = self.generate_ticket_url(timesheet_info)

            if ticket_url:
                self.view.window().status_message(
                    'Opening {}'.format(ticket_url)
                )
                webbrowser.open(ticket_url)
        else:
            self.view.window().status_message(
                "Can't find ticket in current line"
            )

    def is_visible(self, *args) -> bool:
        """
        Called when context menu is appeared.
        Return True if current line has ticket info
        and Goto Ticket menu item should appear.
        Return False otherwise.
        """
        return self.timesheet_helper.is_valid_timesheet_under_cursor()

    def generate_ticket_url(self, timesheet_info: dict) -> Optional[str]:
        """
        Generate and return ticket URL to given `ticket_info`.
        If required setting isn't set, or doesn't have template,
        show error message and return None.
        """
        error_prefix = 'Timesheets plugin'

        bug_tracker, ticket_id = timesheet_info['issue']

        settings = sublime.load_settings('timesheets.sublime-settings')

        option_name = '{}_ticket_url'.format(bug_tracker)
        ticket_url_template = settings.get(option_name)

        if not ticket_url_template:
            sublime.error_message(
                '{}: in order to open ticket in browser '
                'please specify valid "{}" option'.format(
                    error_prefix,
                    option_name
                )
            )
            return

        ticket_url = ticket_url_template.format(ticket_id)

        # If formatted template and raw template are the same,
        # looks like template missing placeholder ("{}")
        if ticket_url == ticket_url_template:
            sublime.error_message(
                '{}: in order to open ticket in browser '
                'please specify valid "{}" option'.format(
                    error_prefix,
                    option_name
                )
            )
            return

        return ticket_url


time_worked.py
import sublime_plugin

from timesheets.helpers import prettify_minutes
from timesheets.helpers import SublimeHelper, TimesheetHelper


class TimeWorked(sublime_plugin.ViewEventListener):
    """
    Display today and week worked time in status bar for opened timesheet file.
    """

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.sublime_helper = SublimeHelper(self.view)
        self.timesheet_helper = TimesheetHelper(self.sublime_helper)

        # Remember if current view contains timesheet file.
        # This needed to save resources and don't process non-timesheet files.
        self.is_timesheet = None

    def on_activated(self):
        """
        Called when file is opened, tab is activated by clicking,
        or when minimized Sublime is restored (e.g. by Alt-Tab).
        """
        if not self.detect_is_timesheet():
            return

        self.update_worked_message()

    def on_post_save(self):
        if not self.detect_is_timesheet():
            return

        self.update_worked_message()

    def detect_is_timesheet(self) -> bool:
        """
        If it's unknown whether content of current view is timesheet or not,
        detect it. Otherwise use saved value.
        In both cases return whether content of current view
        looks like timesheet.
        """
        if self.is_timesheet is None:
            self.is_timesheet = self.timesheet_helper.is_timesheet()
        return self.is_timesheet

    def update_worked_message(self):
        """Calculate and update worked today time in status bar."""
        worked_today_minutes = self.timesheet_helper.worked_today_minutes()
        worked_week_minutes = self.timesheet_helper.worked_week_minutes()

        message = 'Worked today {}, week {}'.format(
            prettify_minutes(worked_today_minutes),
            prettify_minutes(worked_week_minutes),
        )

        self.view.set_status('timesheet', message)


jira_completion.py
"""
Sublime Text 3 Plugin to autocomplete JIRA issues numbers
by ticket titles.

How to use:

Type "JI" with an optional filter, e.g. "JIauthgoogle", press Alt+/.
Completion menu should pop up.

How to set up:
1. Add your username and password to ~/.netrc, e.g.:

    machine jira.iponweb.net login vpetrov password topsecretpassword

CAUTION: async requests are not supported, so your editor will freeze
until reply from JIRA is received.
Also, minimal error handling, sorry.
"""

import base64
import datetime
import json
import netrc
import sublime
import sublime_plugin
import urllib

from urllib.error import HTTPError
from urllib.parse import urlparse


class Settings:
    def __init__(self, settings):
        self._settings = settings

    @property
    def hostname(self):
        return urlparse(self.jira_completion_url).netloc

    @property
    def trigger(self):
        return self._settings.get('trigger', 'JI')

    @property
    def jira_completion_url(self):
        return self._settings.get(
            'jira_completion_url',
            'https://jira.iponweb.net/rest/api/2/search?'
            'jql=(watcher+in+({username})+'
                'OR+assignee+in+({username})+'
                'OR+reporter+in+({username}))+'
                'ORDER+BY+updated+DESC,+created+DESC'
                '&maxResults=100&fields=id,key,summary'
        )

    @property
    def timeout(self):
        return self._settings.get('timeout', 5)

    @property
    def log_filename(self):
        return self._settings.get('log_filename', None)


def load_settings():
    return Settings(sublime.load_settings('timesheets.sublime-settings'))


class JiraCompleter:
    def __init__(self, settings):
        self._settings = settings

    def _log(self, message):
        if not self._settings.log_filename:
            return

        with open(self._settings.log_filename, 'a') as file_:
            file_.write('%s %s\n' % (datetime.datetime.now(), message))

    def _read_credentials(self):
        username, _account, password = \
            netrc.netrc().hosts[self._settings.hostname]
        return username, password

    def _build_url(self):
        username, _password = self._read_credentials()
        return self._settings.jira_completion_url.format(username=username)

    def _get_url(self, url):
        username, password = self._read_credentials()
        opener = urllib.request.build_opener()
        header = base64.b64encode(
            ('%s:%s' % (username, password)).encode('ascii')).decode('ascii')
        opener.addheaders = [('Content-Type', 'application/json')]
        opener.addheaders = [('Authorization', 'Basic %s' % header)]
        try:
            response_body = opener.open(
                url, timeout=self._settings.timeout
            ).read()
        except HTTPError as e:
            self._log('Network error: %s %s' % (e, e.read()))
            raise
        return response_body.decode('utf-8')

    def _get_tickets(self):
        body = self._get_url(self._build_url())
        response = json.loads(body)
        return [
            (entry['key'], entry['fields']['summary'])
            for entry in response['issues']
        ]

    def get_jira_suggestions_online(self):
        self._log('getting online')
        pairs = self._get_tickets()
        suggestions = [
            ('%s %s\t%s' % (self._settings.trigger, subject, key,), key)
            for key, subject in pairs
        ]
        self._log('# suggestions %s' % len(suggestions))
        return suggestions

    def get_suggestions(self):
        try:
            return self.get_jira_suggestions_online()
        except Exception as e:
            self._log('Error getting suggestions: {}'.format(e))
            return []


class JiraAutocomplete(sublime_plugin.EventListener):
    def on_query_completions(self, view, prefix, locations):
        settings = load_settings()

        if not prefix.startswith(settings.trigger):
            return []

        completer = JiraCompleter(settings)
        return completer.get_suggestions()


duplicate_timesheet_line.py
from datetime import datetime, time, timedelta
from time import time as unixtime

import sublime
import sublime_plugin

from timesheets.typing.typing import Optional
from timesheets.helpers import SublimeHelper, TimesheetHelper


class DuplicateTimesheetLineCommand(sublime_plugin.TextCommand):
    """
    Duplicate timesheet line under cursor,
    insert it into appropriate place,
    fill time_from and/or time_to fields.
    """

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.sublime_helper = SublimeHelper(self.view)
        self.timesheet_helper = TimesheetHelper(self.sublime_helper)

        self.regions_to_highlight = []

    def run(self, edit: sublime.Edit, copy_issue_and_comment: bool=True):
        """Entry-point, called when command is executed."""

        # Extract timesheet info from line under cursor
        timesheet_info = self.timesheet_helper.extract_timesheet_info(
            self.sublime_helper.get_current_line_content()
        )

        # Do nothing if current line is not a timesheet line
        if not timesheet_info:
            self.view.window().status_message(
                'No valid timesheet line under cursor to duplicate'
            )
            return

        # Fill time_to if needed,
        # remember it to later fill time_from of new timesheet line
        new_time_to = self.fill_time_to(edit)

        new_time_to = new_time_to or self.floor_time(datetime.now())

        # Get last lines
        last_non_empty_line = self.get_last_non_empty_line_region()
        last_timesheet_line = self.get_last_timesheet_line_region()

        last_timesheet_line_info = self.timesheet_helper.extract_timesheet_info(
            self.view.substr(last_timesheet_line)
        )

        # Determine line after which new timesheet line should be inserted,
        # and if extra new line is needed.
        # Extra new line divides timesheet lines of different days,
        # or acts like separator from latest comment
        if self.timesheet_helper.is_today(last_timesheet_line_info):
            insert_after_line = last_timesheet_line
            extra_newline = False
        else:
            insert_after_line = last_non_empty_line
            extra_newline = True

        # Compose new timesheet line content
        new_timesheet_line_content = '\n{},{},{},{},{}'.format(
            new_time_to.strftime('%Y-%m-%d'),
            new_time_to.strftime('%H:%M'),
            ' ' * len('12:00'),
            timesheet_info['issue'][1] if copy_issue_and_comment else '',
            timesheet_info['comment'] if copy_issue_and_comment else '',
        )

        # Insert extra new line if needed
        if extra_newline:
            self.view.insert(edit, insert_after_line.end(), '\n')
            self.regions_to_highlight.append(sublime.Region(
                insert_after_line.end() + len('\n'),
                insert_after_line.end() + len('\n') + len('\n')
            ))
            insert_after_line = self.view.line(insert_after_line.end() + 1)

        # Insert new line
        self.view.insert(
            edit,
            insert_after_line.end(),
            new_timesheet_line_content
        )

        added_line_region = sublime.Region(
            insert_after_line.end() + len('\n'),
            insert_after_line.end() + len('\n') + len(
                new_timesheet_line_content
            ),
        )

        self.regions_to_highlight.append(added_line_region)

        # Move cursor to new inserted timesheet line,
        # keep column position same as original cursor
        row, column = self.view.rowcol(self.view.sel()[0].begin())
        self.view.sel().clear()
        self.view.sel().add(insert_after_line.end() + len('\n') + column)

        self.view.show(added_line_region)

        self.highlight_regions()

        self.view.window().status_message('Added new timesheet line')

    def is_visible(self, *args, **kwargs):
        """
        Called when context menu is appeared.
        Return True if current line has ticket info
        and Duplicate Timesheet Line menu item should appear.
        Return False otherwise.
        """
        return self.timesheet_helper.is_valid_timesheet_under_cursor()

    def fill_time_to(self, edit: sublime.Edit) -> Optional[datetime]:
        """
        If latest timesheet line is today and has empty time_to field,
        fill it with current time. Otherwise return time_to value.
        If there is no timesheet line, return None.
        """

        last_timesheet_line_region = self.get_last_timesheet_line_region()
        last_timesheet_line_info = self.timesheet_helper.extract_timesheet_info(
            self.view.substr(last_timesheet_line_region)
        )

        # If latest timesheet line isn't today, do nothing
        if not self.timesheet_helper.is_today(last_timesheet_line_info):
            return

        # If time_to already filled, return it
        if last_timesheet_line_info['to_dt']:
            return last_timesheet_line_info['to_dt']

        # Find start position of time_to field (it's fixed)
        time_to_start = len('2000-01-01,12:00,')

        # Find end position of time_to field
        # which could be any number of space chars
        pos = last_timesheet_line_region.begin() + time_to_start
        while pos < last_timesheet_line_region.end():
            if self.view.substr(pos) == ',':
                break
            pos += 1

        new_time_to = self.floor_time(datetime.now())

        # Replace empty time_to with current time
        self.view.replace(
            edit,
            sublime.Region(
                last_timesheet_line_region.begin() + time_to_start,
                pos
            ),
            new_time_to.strftime('%H:%M')
        )

        # Add modified piece of line to highlight
        self.regions_to_highlight.append(sublime.Region(
            last_timesheet_line_region.begin() + time_to_start,
            last_timesheet_line_region.begin() + time_to_start + len('12:00'),
        ))

        return new_time_to

    def highlight_regions(self):
        """
        Highlight new inserted content so user knows what was modified.
        Also remove that highlight after short time so it won't bother user.
        """
        # For each inserted line generate unique key
        regions_key = 'timesheets-{}'.format(unixtime())

        # Highlight regions
        self.view.add_regions(
            regions_key,
            self.regions_to_highlight,
            'text',
            'dot',
            sublime.DRAW_NO_FILL
        )

        self.regions_to_highlight.clear()

        # Schedule regions clear
        sublime.set_timeout(lambda: self.view.erase_regions(regions_key), 750)

    def get_last_non_empty_line_region(self) -> Optional[sublime.Region]:
        """
        Return line region of latest non-empty string (including comments).
        """
        for line_region in self.sublime_helper.iter_lines_regions_reversed():
            line_content = self.view.substr(line_region)

            if line_content.strip():
                return line_region

    def get_last_timesheet_line_region(self) -> Optional[sublime.Region]:
        """Return line region of latest timesheet line (excluding comments)."""
        for line_region in self.sublime_helper.iter_lines_regions_reversed():
            line_content = self.view.substr(line_region)

            if self.timesheet_helper.extract_timesheet_info(line_content):
                return line_region

    def floor_time(self, dt: datetime) -> datetime:
        """
        Return earliest time to given `dt` with rounding to 10 minutes.
        E.g.:
        12:00 -> 12:00
        12:01 -> 12:00
        12:09 -> 12:00
        """
        return datetime.combine(dt.date(), time(dt.hour, dt.minute // 10 * 10))

    def ceil_time(self, dt: datetime) -> datetime:
        """
        Return closest time in future to given `dt` with rounding to 10 minutes.
        E.g.:
        12:00 -> 12:10
        12:01 -> 12:10
        12:09 -> 12:10
        """
        return self.floor_time(dt) + timedelta(minutes=10)


helpers.py
import re
from datetime import datetime

import sublime

from timesheets.typing.typing import Optional, Generator, Iterable, List


class SublimeHelper:
    """Utility class with convenient methods to work with given `view`."""

    def __init__(self, view: sublime.View):
        self.view = view

    def get_current_line_content(self) -> str:
        """
        Return content of line where cursor is placed.
        If there are few cursors, use only first cursor.
        """
        current_selection = self.view.sel()[0]
        line = self.view.line(current_selection)
        return self.view.substr(line)

    def iter_lines(self) -> Generator[str, None, None]:
        """
        Iterate content of view line by line.
        Trailing newline character not included.
        """
        pos = 0
        while pos < self.view.size():
            line_region = self.view.line(pos)
            line_content = self.view.substr(line_region)

            yield line_content

            pos = line_region.end() + 1

    def iter_lines_regions_reversed(self) -> Iterable[sublime.Region]:
        """
        Iterate regions of view line by line in reverse order
        (from last line to first).
        Trailing newline character not included.
        """
        pos = self.view.size()
        while pos >= 0:
            line_region = self.view.line(pos)

            yield line_region

            pos = line_region.begin() - 1

    def iter_lines_reversed(self) -> Generator[str, None, None]:
        """
        Iterate content of view line by line in reverse order
        (from last line to first).
        Trailing newline character not included.
        """
        for line_region in self.iter_lines_regions_reversed():
            line_content = self.view.substr(line_region)

            yield line_content


class TimesheetHelper:
    """
    Utility class with convenient methods to work with timesheet
    from view that's attached to given `sublime_helper`.
    """

    def __init__(self, sublime_helper: SublimeHelper):
        self.sublime_helper = sublime_helper

    def extract_timesheet_info(self, line: str) -> Optional[dict]:
        """
        Try to extract timesheet info from given `line`.
        Return it as dict if content contains timesheet,
        or None otherwise.
        """
        match = re.search(
            r'^(?P<year>\d{4})-(?P<month>\d{2})-(?P<day>\d{2}),'
            r'(?P<from_hour>\d{2}):(?P<from_min>\d{2}),'
            r'(?P<to>\d{2}:\d{2}|[\s]*),'
            r'(?P<jira_issue>[\w_\-\d]+),'
            r'(?P<comment>.*)$',
            line
        )

        if not match:
            return

        try:
            from_dt = datetime(
                int(match.group('year')),
                int(match.group('month')),
                int(match.group('day')),
                int(match.group('from_hour')),
                int(match.group('from_min')),
            )
        except ValueError:
            return

        if match.group('to').strip():
            to_hour, to_min = match.group('to').split(':')

            try:
                to_dt = datetime(
                    int(match.group('year')),
                    int(match.group('month')),
                    int(match.group('day')),
                    int(to_hour),
                    int(to_min),
                )
            except ValueError:
                return
        else:
            to_dt = None

        issue = ('jira', match.group('jira_issue'))

        return {
            'from_dt': from_dt,
            'to_dt': to_dt,
            'issue': issue,
            'comment': match.group('comment'),
        }

    def is_timesheet(self) -> bool:
        """
        Return True if current file is timesheet, False otherwise.
        Very simple heuristic is used to determine it:
        if more than 90% of first 100 lines are valid timesheet lines
        (except empty lines and comments), then file considered timesheet.
        """
        valid_timesheet_lines = 0
        invalid_timesheet_lines = 0
        for count, line in enumerate(self.sublime_helper.iter_lines()):
            if count >= 100:
                break

            if not line.strip():
                continue

            if self.is_comment(line):
                continue

            if self.extract_timesheet_info(line):
                valid_timesheet_lines += 1
            else:
                invalid_timesheet_lines += 1

        if not valid_timesheet_lines:
            return False

        valid_to_invalid_lines_ratio = (
            valid_timesheet_lines /
            (valid_timesheet_lines + invalid_timesheet_lines)
        )

        return valid_to_invalid_lines_ratio >= 0.9

    def is_today(self, timesheet_info: dict) -> bool:
        """Return true if given `timesheet_info` relates to today."""
        return \
            timesheet_info and \
            timesheet_info['from_dt'].date() == datetime.now().date()

    def worked_today_minutes(self) -> float:
        """Iterate lines backward and count how much time is worked today."""
        today_timesheet_info = []

        for line in self.sublime_helper.iter_lines_reversed():
            timesheet_info = self.extract_timesheet_info(line)
            if not timesheet_info:
                continue

            if timesheet_info['from_dt'].date() == datetime.today().date():
                today_timesheet_info.append(timesheet_info)
            else:
                break

        return self.timesheets_info_to_minutes(today_timesheet_info)

    def worked_week_minutes(self) -> float:
        """
        Iterate lines backward and count how much time is worked this week.
        """
        week_timesheet_info = []

        today_year, today_week, _ = datetime.today().isocalendar()

        for line in self.sublime_helper.iter_lines_reversed():
            timesheet_info = self.extract_timesheet_info(line)
            if not timesheet_info:
                continue

            line_year, line_week, _ = timesheet_info['from_dt'].isocalendar()

            if line_year == today_year and line_week == today_week:
                week_timesheet_info.append(timesheet_info)
            else:
                break

        return self.timesheets_info_to_minutes(week_timesheet_info)

    def timesheets_info_to_minutes(self, timesheets_info:  List[dict]) -> float:
        """
        Return how much time is worked based on given list
        of timesheet info objects.
        """
        worked_minutes = 0
        for timesheet_info in timesheets_info:
            to_dt = timesheet_info['to_dt'] or datetime.now()
            timesheet_line_worked_minutes = (
                to_dt - timesheet_info['from_dt']
            ).total_seconds() / 60
            worked_minutes += timesheet_line_worked_minutes

        return worked_minutes

    def is_comment(self, line: str) -> bool:
        """Return True if given `line` is comment."""
        return line.lstrip().startswith('#')

    def is_valid_timesheet_under_cursor(self) -> bool:
        """Return True if line under cursor is valid timesheet line."""
        return bool(
            self.extract_timesheet_info(
                self.sublime_helper.get_current_line_content()
            )
        )


def prettify_minutes(minutes_total: float) -> str:
    """
    Return time in format "<hours>:<minutes>" from given `minutes_total`.
    E.g. 65 minutes -> "01:05".
    """
    hours, minutes = divmod(minutes_total, 60)
    return '{:02}:{:02}'.format(int(hours), int(minutes))
